contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,267 @@
+    };
+    match msg {
+        StakeChangedHookMsg::Stake { addr, .. } => execute_stake(deps, env, addr),
+        StakeChangedHookMsg::Unstake { addr, .. } => execute_unstake(deps, env, addr),
+    }
+}
+
+pub fn execute_stake(
+    mut deps: DepsMut,
+    env: Env,
+    addr: Addr
+) -> Result<Response<Empty>, ContractError> {
+    update_rewards(&mut deps, &env, &addr)?;
+    Ok(Response::new().add_attribute("action", "stake"))
+}
+
+pub fn execute_unstake(
+    mut deps: DepsMut,
+    env: Env,
+    addr: Addr
+) -> Result<Response<Empty>, ContractError> {
+    update_rewards(&mut deps, &env, &addr)?;
+    Ok(Response::new().add_attribute("action", "unstake"))
+}
+
+pub fn execute_claim(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response<Empty>, ContractError> {
+    update_rewards(&mut deps, &env, &info.sender);
+    let rewards = PENDING_REWARDS
+        .load(deps.storage, info.sender.clone())
+        .map_err(|_| NoRewardsClaimable {})?;
+    PENDING_REWARDS.save(deps.storage, info.sender.clone(), &Uint128::zero());
+    let config = CONFIG.load(deps.storage)?;
+    let transfer_msg = get_transfer_msg(info.sender, rewards, config.reward_token)?;
+    Ok(Response::new().add_message(transfer_msg).add_attribute("action", "claim").add_attribute("amount", rewards))
+}
+
+pub fn get_transfer_msg(recipient: Addr, amount: Uint128, denom: Denom) -> StdResult<CosmosMsg> {
+    match denom {
+        Denom::Native(denom) => {
+            Ok(
+               BankMsg::Send {
+                   to_address: recipient.into_string(),
+                   amount: vec![Coin{ denom: denom, amount }]
+               }.into()
+            )
+        }
+        Denom::Cw20(addr) => {
+            let cw20_msg = to_binary(&cw20::Cw20ExecuteMsg::Transfer {
+                recipient: recipient.into_string(),
+                amount,
+            })?;
+            Ok(WasmMsg::Execute {
+                contract_addr: addr.into_string(),
+                msg: cw20_msg,
+                funds: vec![],
+            }.into())
+        }
+    }
+}
+
+pub fn update_rewards(deps: &mut DepsMut, env: &Env, addr: &Addr) -> StdResult<()> {
+    let config = CONFIG.load(deps.storage)?;
+    let reward_config = REWARD_CONFIG.load(deps.storage)?;
+    let total_staked = get_total_staked(deps.as_ref(), &config.staking_contract)?;
+    let staked_balance = get_staked_balance(deps.as_ref(), &config.staking_contract, addr)?;
+    let current_block = min(env.block.height, reward_config.periodFinish);
+    let last_update_block = LAST_UPDATE_BLOCK.load(deps.storage).unwrap_or_default();
+
+    let reward_per_token = REWARD_PER_TOKEN.update::<_, StdError>(deps.storage, |r| {
+        let additional_reward_per_token = match total_staked {
+            Uint128::zero() => Uint128::zero(),
+            _ => ((reward_config.rewardRate * max(Uint128::from(current_block - last_update_block), Uint128::zero()))
```

Comment:
Is this `max` call needed? A uint should never be smaller than zero.

Reply:
is that how overflow works? will it just stop at 0?

Reply:
Oh I see. Are you worried here about an overflow when you do `current_block - last_update_block`? I guess that is possible if the rewards time is expired and someone has caused `LAST_UPDATE_BLOCK` to be updated.

You probably just want something like:

```rust
if current_block < last_update_block {
  Uint128::zero()
} {
  Uint128::new(current_block - last_update_block)
}
```

Rusts overflow behavior is to wrap. So something like `0u32 - 1` would wrap around to `2^32`. Taking the max here won't help as you'll just get the overflowed value. [This](https://huonw.github.io/blog/2016/04/myths-and-legends-about-integer-overflow-in-rust/) blog post has some nice info about that.

Reply:
ah lol, good catch, will implement better overflow handling

Reply:
this is resolved, as last update block will only get updated to the end of the current rewards period